### PR TITLE
Cache browser visibility state & emit visibilitychange event on change

### LINKED ISSUE #3788: visibilitychange event not being triggered
I've been trying to use the [Page Visibility API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) however the `visibilitychange` event is not being triggered. 

I understand that Electron's `pageVisibility` option forces either visible or hidden states but is there a way to listen for the true window visibility changes?

P.S.: Here's a small snippet and an [electron-quick-start fork](https://github.com/zenorocha/electron-quick-start/blob/master/index.html) if you guys wanna do some testing.

```js
document.addEventListener("visibilitychange", function() {
  console.log( document.visibilityState );
});
```

### PR DESCRIPTION
**Pull request by ArekSredzki**
Fixes #3788 & #4669

### CODE REVIEW

**Comment by ArekSredzki** [comment]
Can anyone comment on why those two builds failed? They did so silently... weird.

**Comment by KevinMartin** [comment]
It could be because you changed many of the comparators from from abstract equality (`==`) to strict equality (`===`). Some of those values can be undefined, yet now you're checking solely for `null`. 

If this PR solely focuses on browser visibility state, it shouldn't be changing the code style of miscellaneous files.

Try reverting those changes and submit only the changes necessary.

**Comment by ArekSredzki** [comment]
@KevinMartin Thanks, I did so, but the result is the same. Do the tests normally start to run immediately?

File: atom/browser/api/lib/browser-window.js [hunk 1]
```diff
@@ -88,6 +88,18 @@ BrowserWindow.prototype._init = function() {
     };
   })(this));
 
+  // Evented visibilityState changes
+  this.on('minimize', (function(_this) {
```
Comment (zcbenz) [inline atom/browser/api/lib/browser-window.js]: You can use ES6 arrow function.

**Comment by zcbenz** [comment]
The code also needs to handle the case when the window show/hides.

**Comment by ArekSredzki** [comment]
@zcbenz There is currently no event emitted when those two things happen. I'm guessing that I should add those?

**Comment by ArekSredzki** [comment]
@KevinMartin @zcbenz The CI build is failing because I rebased from master, which is failing, but this PR is ready to be merged in.

**Comment by ArekSredzki** [comment]
@KevinMartin @zcbenz This PR is now ready to be merged into mainline. There seems to be a lot of jitter in your builds though... The test that timed-out shouldn't have, and hasn't in the past.
I'd really appreciate having this be included in the next electron release, let me know if there's anything that's blocking that from happening.

Thanks.

**Comment by ArekSredzki** [comment]
@zcbenz ... Any chance you could respond...

**Comment by phamdaniel** [comment]
You're going to have to fix the merge conflicts as `.js` files have been moved out of the `atom/` folder.
Also I'd say to be patient as there's a lot of important work happening to get Electron up to Chrome 49

**Comment by ArekSredzki** [comment]
@phamdaniel Done

Sorry about coming off as impatient. Just don't want this PR to constantly have to keep updating it to maintain compatibility with master and didn't see why other PRs were being merged without hesitation.

**Comment by zcbenz** [comment]
:+1: 

Sorry we are in the process of upgrading to Chrome 49, so reviews of pull requests can be delayed.

**Pull request merged**